File: apps/web/src/types/testExecution.ts
export interface TestExecutionRun {
  id: number;
  run_code: string;
  name: string;
}

export interface TestExecutionTestCase {
  id: number;
  test_case_code: string;
  title: string;
  priority: string;
  preconditions: string | null;
  steps: string | null;
  expected_result: string | null;
}

export interface TestExecution {
  id: number;
  run_id: number;
  test_case_id: number;
  status: string;
  actual_result: string | null;
  comments: string | null;
  executed_by: string | null;
  executed_at: string | null;

  test_run: TestExecutionRun;
  test_case: TestExecutionTestCase;

  created_at: string;
  updated_at: string;
}

export interface TestExecutionSummary {
  total: number;
  passed: number;
  failed: number;
  blocked: number;
  not_executed: number;
  pass_percentage: number;
}

export interface TestExecutionUpdateRequest {
  run_id: number;
  test_case_id: number;
  status: string;
  actual_result: string | null;
  comments: string | null;
  executed_by: string | null;
  executed_at: string | null;
}

File: apps/api/app/schemas/test_execution.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel


class TestExecutionRunResponse(BaseModel):
    id: int
    run_code: str
    name: str

    class Config:
        from_attributes = True


class TestExecutionTestCaseResponse(BaseModel):
    id: int
    test_case_code: str
    title: str
    priority: str
    preconditions: str | None = None
    steps: str | None = None
    expected_result: str | None = None


    class Config:
        from_attributes = True


class TestExecutionBase(BaseModel):
    run_id: int
    test_case_id: int
    status: str = "Not Executed"
    actual_result: Optional[str] = None
    comments: Optional[str] = None
    executed_by: Optional[str] = None
    executed_at: Optional[datetime] = None


class TestExecutionCreate(TestExecutionBase):
    pass


class TestExecutionUpdate(TestExecutionBase):
    pass


class TestExecutionResponse(TestExecutionBase):
    id: int
    test_run: TestExecutionRunResponse
    test_case: TestExecutionTestCaseResponse
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True